go/pkg/reader/reader.go

```diff
@@ -107,3 +111,75 @@ func (fio *fileSeeker) Initialize() error {
 	fio.initialized = true
 	return nil
 }
+
+type compressedSeeker struct {
+	fs   ReadSeeker
+	encd *zstd.Encoder
+	// This keeps the compressed data
+	buf *bytes.Buffer
+}
+
+func NewCompressedFileSeeker(path string, buffsize int) (ReadSeeker, error) {
+	return NewCompressedSeeker(NewFileReadSeeker(path, buffsize))
+}
+
+func NewCompressedSeeker(fs ReadSeeker) (ReadSeeker, error) {
+	if _, ok := fs.(*compressedSeeker); ok {
+		return nil, errors.New("Trying to double compress files.")
+	}
+
+	buf := bytes.NewBuffer(nil)
+	encd, err := zstd.NewWriter(buf)
+	return &compressedSeeker{
+		fs:   fs,
+		encd: encd,
+		buf:  buf,
+	}, err
+}
+
+func (cfs *compressedSeeker) Read(p []byte) (int, error) {
+	var n int
+	var errR, errW error
+	for cfs.buf.Len() < len(p) && errR == nil && errW == nil {
```

Comment:
Re: errors: I meant something like:

```
defer cfs.encd.Close()
for cfs.buf.Len() < len(p) {
  n, err := cfs.fs.Read(p)
  if err != nil {
    return 0, err
  }
  if _, err = cfs.encd.Write(p[:n]); err != nil {
    return 0, err
  }
}
return cfs.buf.Read(p)
```

Wouldn't that be equivalent? Or am I missing something?

Reply:
And if you really need to close the encoder conditionally on no previous errors, it should be doable with:

```
func (cfs *compressedSeeker) Read(p []byte) (n int, err error) {
  defer func() {
    if err != nil {
      cfs.encd.Close()
     }
  }()
...
}
```

Reply:
The first block is different whereas in I'm still doing the Write after the loop Read, regardless if the loop Read returned an error. Basically I'm making sure that I'm compressing *all* the data I have available. Besides, I like the code having fewer indentation.

Additionally, I want to Close the enconder on an *existing* error (ie EOF or any other error that'd make continuing the compression impossible). It would be better to use the defer if I was preemptively returning errors in your first suggestion - otherwise it merely adds indentation.